open_spiel/games/dark_chess.cc

```diff
@@ -0,0 +1,363 @@
+}
+
+// Adds a plane to the information state vector corresponding to the presence
+// and absence of the given piece type and colour at each square.
+void AddPieceTypePlane(chess::Color color, chess::PieceType piece_type,
+                       const chess::ChessBoard& board,
+                       std::array<bool, chess::kMaxBoardSize * chess::kMaxBoardSize>& observability_table,
+                       absl::Span<float>::iterator& value_it) {
+  for (int8_t y = 0; y < chess::kMaxBoardSize; ++y) {
+    for (int8_t x = 0; x < chess::kMaxBoardSize; ++x) {
+      auto square = chess::Square{x, y};
+      chess::Piece piece_on_board = board.at(square);
+      *value_it++ =
+          (piece_on_board.color == color && piece_on_board.type == piece_type && observability_table[chess::SquareToIndex(square, board.BoardSize())]
+           ? 1.0
+           : 0.0);
+    }
+  }
+}
+
+// Adds a plane to the information state vector corresponding to the observability of given square
+void AddUnknownSquarePlane(std::array<bool, chess::kMaxBoardSize * chess::kMaxBoardSize>& observability_table,
+                           absl::Span<float>::iterator& value_it) {
+  for (int i = 0; i < chess::kMaxBoardSize * chess::kMaxBoardSize; ++i) {
+    *value_it++ = observability_table[i] ? 0.0 : 1.0;
+  }
+}
+
+// Adds a uniform scalar plane scaled with min and max.
+template <typename T>
+void AddScalarPlane(T val, T min, T max,
+                    absl::Span<float>::iterator& value_it) {
+  double normalized_val = static_cast<double>(val - min) / (max - min);
+  for (int i = 0; i < chess::kMaxBoardSize * chess::kMaxBoardSize; ++i)
+    *value_it++ = normalized_val;
+}
+
+// Adds a binary scalar plane.
+void AddBinaryPlane(bool val, absl::Span<float>::iterator& value_it) {
+  AddScalarPlane<int>(val ? 1 : 0, 0, 1, value_it);
+}
+
+}  // namespace
+
+
+DarkChessState::DarkChessState(std::shared_ptr<const Game> game, int boardSize, const std::string& fen)
+    : State(game),
+      start_board_(*chess::ChessBoard::BoardFromFEN(fen, boardSize, true)),
+      current_board_(start_board_) {
+  SPIEL_CHECK_TRUE(&current_board_);
+  repetitions_[current_board_.HashValue()] = 1;
+
+  aohs_.reserve(2);
+  for (Player player = 0; player < NumPlayers(); ++player) {
+    std::vector<std::pair<std::optional<Action>, std::string>> aoh;
+    aoh.push_back({static_cast<std::optional<Action>>(std::nullopt), ObservationString(player)});
+    aohs_.push_back(open_spiel::ActionObservationHistory(player, aoh));
+  }
+}
+
+void DarkChessState::DoApplyAction(Action action) {
+  Player current_player = CurrentPlayer();
+  chess::Move move = ActionToMove(action, Board());
+  moves_history_.push_back(move);
+  Board().ApplyMove(move);
+  ++repetitions_[current_board_.HashValue()];
+  cached_legal_actions_.reset();
+
+  for (Player player = 0; player < NumPlayers(); ++player) {
+    auto a = current_player == player ? action : static_cast<std::optional<Action>>(std::nullopt);
```

Comment:
static_cast<std::optional<Action>>(std::nullopt) -- sometimes can be replaced with {} if the compiler can deduce the type.